feat: optionally install Scenario with `ops[testing]` and expose the names in ops.testing (#1381)
canonical/operator: docs/index.rst

The library (`available on PyPI`_) provides:

- :ref:`ops_module`, the API to respond to Juju events and manage the application;
- :ref:`ops_main_entry_point`, used to initialise and run your charm;
- :ref:`ops_pebble_module`, the Pebble client, a low-level API for Kubernetes containers;
- the APIs for unit testing charms in a simulated environment:

  - :doc:`State-transition testing </state-transition-testing>`. This is the

>     recommended approach.

Reviewer: ```suggestion
    recommended approach (it was previously known as 'Scenario').
```

  - :doc:`Harness </harness>`. This is a deprecated framework, and has issues,
    particularly with resetting the charm state between Juju events. It will be

>     moved out of the base ``ops`` install in an ops release in the future. Charm

Reviewer: ```suggestion
    moved out of the base ``ops`` install in an ops release in the future; at that point, charm
```

Reviewer: It also seems like too much information for this point in the docs. Can we move everything from "It will be moved out..." to the Harness page?

Author: Happy to move it - in terms of "at that point": I updated the description to include this but didn't mention it in a comment: I added the ability to do `pip install ops[harness]` now (this was in the spec, but I had forgotten). This eases transition in the future, because charms can use `pip install ops[harness]` now and it should seamlessly work when it becomes separated.

ops.testing module
==================

Frameworks for unit testing charms in a simulated Juju environment.

Two frameworks are available:

* State-transition testing, which tests the charm's state transitions in response
  to events. This is the recommended approach. Install ops with the ``testing``
  extra to use this framework; for example: ``pip install ops[testing]``
* Harness, which provides an API similar to the Juju CLI. This is a legacy
  framework, and has issues, particularly with resetting the charm state between
  Juju events. This is currently included with a base ``ops`` install.

.. note::
    Unit testing is only one aspect of a comprehensive testing strategy. For more
    on testing charms, see `Charm SDK | Testing <https://juju.is/docs/sdk/testing>`_.

> State-transition testing

Author: I've split it out - you might want to take another look now?

------------------------

> Write tests that declaratively define the Juju state all at once, define the

Reviewer: This reads a bit odd, like a subject is missing. I think we should say "State-transition tests allow you to ..." or something.

Juju context against which to test the charm, and fire a single event on the
charm to execute its logic. The tests can then assert that the Juju state has

> changed as expected.

Reviewer: There's a lot of text after this point before the developer can see anything in action. I think it'd be great to show a tiny example right here:

```python
def test_start():
    ctx = Context(MyCharm)
    state = State(...something small here...)
    out = ctx.run(ctx.on.start(), state)
    assert isinstance(out.unit_status, ActiveStatus)
```

Maybe we want *something* really simple in the `State` so readers can see that.

> These tests are 'state-transition' tests, a way to test isolated units of charm
> functionality (how the state changes in reaction to events). They are not
> necessarily tests of individual methods or functions (but might be, depending on
> the charm's event observers); they are testing the 'contract' of the charm: given
> a certain state, when a certain event happens, the charm should transition to a
> certain (likely different) state. They do not test against a real Juju
> controller and model, and focus on a single Juju unit, unlike integration tests.
> For simplicity, we refer to them as 'unit' tests in the charm context.

Reviewer: ```suggestion
These 'state-transition' tests give charm authors a way to test
how the state changes in reaction to events. They are not
necessarily tests of individual methods or functions;
they are testing the 'contract' of the charm: given
a certain state, when a certain event happens, the charm should transition to
another state. Unlike integration tests, they do not test using a real Juju
controller and model, and focus on a single Juju unit.
For simplicity, we refer to them as 'unit' tests.
```

> Writing these tests should nudge you into thinking of a charm as a black-box
> input->output function. The input is the union of an `Event` (why am I, charm,
> being executed), a `State` (am I leader? what is my integration data? what is my
> config?...) and the charm's execution `Context` (what integrations can I have?
> what containers can I have?...). The output is another `State`: the state after
> the charm has had a chance to interact with the mocked Juju model and affect the
> state.

Reviewer: ```suggestion
Writing these tests should nudge you into thinking of a charm as a black-box
'input to output' function. The inputs are:

- Event: why am I, the charm, being executed
- State: am I the leader? what is my integration data? what is my config?
- Context: what integrations can I have? what containers can I have?

The output is another `State`: the state after
the charm has interacted with the mocked Juju model.
The output state is the same type of data structure as the input state.
```

.. image:: https://raw.githubusercontent.com/canonical/ops-scenario/main/resources/state-transition-model.png
   :alt: Transition diagram, with the input state and event on the left, the context including the charm in the centre, and the state out on the right

Writing unit tests for a charm, then, means verifying that:

- the charm does not raise uncaught exceptions while handling the event

> - the output state (as compared with the input state) is as expected.

Reviewer: I would reverse those two bullet points as the "output state" one is by far the most important.

A test consists of three broad steps:

- **Arrange**:
    - declare the context
    - declare the input state
- **Act**:
    - select an event to fire

>     - run the context (i.e. obtain the output state, given the input state and the event)

Reviewer: "run the context" sounds odd. Maybe this could just be one bullet point with "run a specific Juju event"?

- **Assert**:

>     - verify that the output state (as compared with the input state) is how you expect it to be

Reviewer: ```suggestion
    - verify that the output state is what you expect it to be
```

    - verify that the charm has seen a certain sequence of statuses, events, and `juju-log` calls

>     - optionally, you can use a context manager to get a hold of the charm instance and run

Reviewer: Maybe drop this bullet point altogether? We don't need to mention this more advanced feature right here.

      assertions on APIs and state internal to it.

> The most basic scenario is one in which all is defaulted and barely any data is

Reviewer: The wording here is a bad odd/casual. But maybe if we start with a tiny example above, we don't need this paragraph or example at all?

available. The charm has no config, no integrations, no leadership, and its
status is `unknown`. With that, we can write the simplest possible test:

.. code-block:: python

    def test_base():
        ctx = Context(MyCharm)
        out = ctx.run(ctx.on.start(), State())
        assert out.unit_status == UnknownStatus()

> .. autoclass:: ops.testing.ActionFailed

Reviewer: It's a pity we have to list these out manually? Is there any way to avoid that, so that we don't forget to add new classes here when we add them to the code?

Author: I do agree. I couldn't get `automodule` to work and also use the `ops.testing` name rather than leaking `ops._private.harness` or `scenario`. It's somewhat working against the `automodule` intention since the usual desired behaviour is to document anything defined in the module but not anything that the module pulls in to use - but here we have basically nothing defined in the module, and want to document (almost) everything that's pulled in.

There could be some combination of Sphinx configuration options that make it work, but I wasn't able to find them with the limited amount of time I allowed for it. It's also possible that we could create our own `automodule` directive (subclassing or patching the original one, probably) that would solve our issues - but this seemed like it would be a lot more work.

For now, I've added a test_infra test to make it less likely that we forget to maintain this, and a comment about why we do this, and hopefully someone in the future will make this nicer.
